indexing-hadoop/src/main/java/io/druid/indexer/InputRowSerde.java

```diff
@@ -123,24 +98,73 @@ public InputRow get()
         String t = aggFactory.getTypeName();
 
         if (t.equals("float")) {
-          mw.put(new Text(k), new FloatWritable(agg.getFloat()));
+          out.writeFloat(agg.getFloat());
         } else if (t.equals("long")) {
-          mw.put(new Text(k), new LongWritable(agg.getLong()));
+          WritableUtils.writeVLong(out, agg.getLong());
         } else {
           //its a complex metric
           Object val = agg.get();
           ComplexMetricSerde serde = getComplexMetricSerde(t);
-          mw.put(new Text(k), new BytesWritable(serde.toBytes(val)));
+          writeBytes(serde.toBytes(val), out);
         }
       }
 
-      mw.write(out);
       return out.toByteArray();
     } catch(IOException ex) {
       throw Throwables.propagate(ex);
     }
   }
 
+  private static void writeBytes(byte[] value, ByteArrayDataOutput out) throws IOException
+  {
+    WritableUtils.writeVInt(out, value.length);
+    out.write(value, 0, value.length);
+  }
+
+  private static void writeString(String value, ByteArrayDataOutput out) throws IOException
+  {
+    writeBytes(value.getBytes(Charsets.UTF_8), out);
+  }
+
+  private static void writeStringArray(List<String> values, ByteArrayDataOutput out) throws IOException
+  {
+    if(values == null || values.size() == 0) {
+      WritableUtils.writeVInt(out, 0);
```

Comment:
Yes, this lines all are needed, because it change the data format communicate between map and reduce, and they all just used between map and reduce to deliver data not any where.